Add  missing explicit error expectation for transaction mutations

ffakenz opened the pull request:
This contributes to moving #705 forward.

To check before merging:
* [x] CHANGELOG is up to date
* [x] Up to date with master

ffakenz committed:
Add explicit error expectations for Close/Contest

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Abort.hs:
```diff
@@ -145,26 +145,28 @@ data AbortMutation
-    , SomeMutation Nothing DropCollectedInput . RemoveInput <$> elements (txIns' tx)
+    , SomeMutation (Just "burnt token number mismatch") MutateThreadTokenQuantity <$> changeMintedValueQuantityFrom tx (-1)
+    , SomeMutation (Just "burnt token number mismatch") BurnOneTokenMore <$> addPTWithQuantity tx (-1)
+    , -- REVIEW: why here are we getting back a trace from HeadTokens?
```
This is fine and expected. The minting (HeadTokens) is the one which actually checks the number of tokens being burned (right now). 

In the abort transactions up to 4 scripts run! Head, init, commit and the minting policy script.

ch1bo replied:
In fact, our head token script does check too much. Or things which are redundant. According to the spec we do not need to do these checks in the minting policy as the v_head is already doing it.

Let's address this as part of #720

ffakenz replied:
Outdated.

Addressed as part of #720.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -271,7 +271,8 @@ genCloseMutation (tx, _utxo) =
             , ChangeOutput 0 $ changeHeadOutputDatum (replaceContestationDeadline adjustedContestationDeadline) headTxOut
             ]
-    , SomeMutation Nothing MutateHeadId <$> do
+    , -- REVIEW: There is no trace for this error. Fails due to `unListData`
```
Seems like `unListData` is a builtin. This is definitely unexpected and we should mutate this in a way it does not blow up.

ch1bo replied:
Could also be that we just need to access the value differently in the script. Can't say right now.

ffakenz replied:
Solved by [commit](https://github.com/input-output-hk/hydra/pull/719/commits/620df63558fe1d769837d0f01de582ce5946c278).

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -189,13 +189,13 @@ genContestMutation
               { signature = toPlutusSignatures mutatedSignature
               }
-      , SomeMutation Nothing MutateToNonNewerSnapshot <$> do
+      , SomeMutation (Just "signature verification failed") MutateToNonNewerSnapshot <$> do
```
This is not what we expect as we actually give a `healthySignature (fromInteger mutatedSnapshotNumber)`

ffakenz replied:
Solved by [commit](https://github.com/input-output-hk/hydra/pull/719/commits/fd30892a1f4e840102664dc7c430a2fd2610744e).

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -208,24 +208,25 @@ genContestMutation
           newSigner <- verificationKeyHash <$> genVerificationKey
           pure $ ChangeRequiredSigners [newSigner]
-      , SomeMutation Nothing MutateContestUTxOHash . ChangeOutput 0 <$> do
+      , SomeMutation (Just "too old snapshot") MutateContestUTxOHash . ChangeOutput 0 <$> do
```
This should also fail with the signature not being consistent with the target utxo hash.

ffakenz replied:
Solved by [commit](https://github.com/input-output-hk/hydra/pull/719/commits/67fa210724777f42989e3fd1f60e9036ca0be143).

ch1bo reviewed (commented):
Let's clarify some of the failures further.

ffakenz committed:
Add missing expected error output

ffakenz committed:
Remove redundant error code H11 in favor of H35

To make close checks consistent with contest's

ffakenz committed:
Fix mutation to fail for the good reason: TooOldSnapshot

The validator checks that: contestSnapshotNumber > closedSnapshotNumber

Before we were mutating the closedSnapshotNumber for a value lower than
healthyClosedSnapshotNumber, but we needed to have it bigger than
healthyContestSnapshotNumber

ffakenz committed:
Fix mutation to fail for the good reason: SignatureVerificationFailed

By using 'const $ healthyClosedState' we are producing an output datum state with snapshotNumber = 3,
and because the input datum snapshotNumber = 3, then it fails for the wrong reason TooOldSnapshot,
as it needs to be greater to verify: closedSnapshotNumber' > closedSnapshotNumber.

ffakenz committed:
Improve mutate close utxo hash to avoid pontential flakyness

ffakenz committed:
Avoid checking if ST is present in output during check contest.

This is redundant because is enough to check that the value is preserved,
as we are comming from a valid Closed state having already checked that the ST is present.

This is a breaking change so we upgrade hydra-plutus to 0.10.0.

Update changelog.

ffakenz committed:
Fix mutation to fail for the good reason: SignerIsNotAParticipant

We were changing input using the closed state as script data instad of a contest redeemer.

ffakenz committed:
Minor improvement over the generator of initial close mutation

To be explicit about the transactions being involved.

ffakenz committed:
Fix mutation to fail for the good reason: SignerIsNotAParticipant

We were changing input using the open state as script data instad of a close redeemer.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -434,7 +432,7 @@ checkContest ctx contestationDeadline contestationPeriod parties closedSnapshotN
       contesters' == contester : contesters
 
-  (contestSnapshotNumber, contestUtxoHash, parties', contestationDeadline', contestationPeriod', headId', contesters') =
+  (closedSnapshotNumber', closedUtxoHash', parties', contestationDeadline', contestationPeriod', headId', contesters') =
```
These names were better before IMO. It's the snapshot number and utxo hash of the contest. In statements above, it's more clear if we can say that `contestSnapshotNumber > closedSnapshotNumber`. Which is intuitive, mostly english and can be reviewed easily.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -671,7 +668,6 @@ instance ToErrorCode HeadError where
     MissingCommits -> "H08"
     HeadValueIsNotPreserved -> "H09"
     HasBoundedValidityCheckFailed -> "H10"
-    InvalidSnapshotSignature -> "H11"
```
This is exactly why we would need the template haskell replacing, so we can keep the numbers not scattered around. Or maybe, numbering was a bad idea from me.

ffakenz replied:
Yeah, I know but, do want me to fix the numbering in this PR?

I left this behind for that reason, thinking it would be better to fix it as part of the PR to add TH.

wdyt?

ch1bo replied:
Yes, that's okay

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -313,8 +313,7 @@ checkClose ctx parties initialUtxoHash sig cperiod headPolicyId =
   checkSnapshot
     | closedSnapshotNumber > 0 =
-      traceIfFalse "H11" $
-        verifySnapshotSignature parties closedSnapshotNumber closedUtxoHash sig
```
BTW: This would have been already a breaking change to the scripts. Thought I mention it so you can build intuition.

ffakenz replied:
updating changelog.

ffakenz committed:
Minor code simplification to make contest consistent with close

ffakenz committed:
Add first round of label descriptions for close mutations

ffakenz committed:
Add first round of label descriptions for contest mutations

ffakenz committed:
Add mutation to check snapshot signature if output snapshot number changes

ffakenz committed:
Simplify arbitrary condition to also be consistent with similar mutation for contest

ffakenz committed:
Rename constant reference in close mutations to be consistent with contest

ffakenz committed:
Improve first round on close mutations to be consistent with contest

ffakenz committed:
Revert on-chain changes.

Notes added to apply them in another PR.

ch1bo committed:
Update some haddocks and rename a Close mutation

ffakenz committed:
Minor changes to contest mutatios to be consistent with close

ffakenz committed:
Minor rewording to simplify descriptions.

We want to enforce the why for a mutation to relevant and not the what is doing.

So here we are removing details about the 'what' is failing about the tx.

We think that is better to follow a white box approach on mutations,
thus labels and error codes should be enough to describe the what.

ffakenz committed:
Add first round of label descriptions for commit mutations

ch1bo committed:
Update haddocks and renamed one mutation label

Break haddock lines on 80 characters makes it easier to read on small or
split screens.

ch1bo committed:
Scale down generated value to pass MutateCommittedValue

On each success, quickcheck increases the size parameter and this will
make generators create bigger values (if they honor the size).

The genValue generator was creating too big values to pass evaluation
within the budget, so we scale them down by a factor of 2.

ffakenz committed:
Add missing mutation description for commit

ffakenz committed:
Add minor missing mutations to close and contest to improve coverage.

ffakenz committed:
Add note to create an error code for a check which is missing a trace in checkClose.

ffakenz committed:
Scale down generated value to pass MutateCommitOutputValue.

On each success, quickcheck increases the size parameter and this will
make generators create bigger values (if they honor the size).

The genValue generator was creating too big values to pass evaluation
within the budget, so we scale them down by a factor of 2.

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -223,29 +227,75 @@ healthyClosedUTxO =
+    -- changing the parties in the input head datum. If they do not align the
+    -- multisignature will not be valid anymore.
+    SnapshotNotSignedByAllParties
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
```
```suggestion
  | -- | Ensures close is authenticated by a one of the Head parties by changing the signer
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -223,29 +227,75 @@ healthyClosedUTxO =
+    -- multisignature will not be valid anymore.
+    SnapshotNotSignedByAllParties
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
+    -- used on the tx to be not one of PTs.
```
```suggestion
    -- used on the tx to not be one of PTs.
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -223,29 +227,75 @@ healthyClosedUTxO =
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
+    -- used on the tx to be not one of PTs.
+    MutateRequiredSigner
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
```
```suggestion
  | -- | Ensures close is authenticated by one of the Head parties by changing the signer
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -223,29 +227,75 @@ healthyClosedUTxO =
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
+    -- used on the tx to have multiple signers (including the signer to not fail for
+    -- SignerIsNotAParticipant).
+    MutateMultipleRequiredSigner
```
@ch1bo: what would be the problem if several head members have signed the close transaction?

ch1bo replied:
No problem. It's just that our current code does not allow it. It's very conservative.

pgrange replied:
ok thanks... I guess we can keep it that way for now then 🤷

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -254,13 +304,13 @@ genCloseMutation (tx, _utxo) =
         pure $ ChangeOutput 0 $ changeHeadOutputDatum (replaceSnapshotNumber mutatedSnapshotNumber) headTxOut
     , SomeMutation (Just $ toErrorCode InvalidSnapshotSignature) MutateSnapshotNumberButNotSignature <$> do
-        mutatedSnapshotNumber <- arbitrarySizedNatural `suchThat` (\n -> n /= healthySnapshotNumber && n > 0)
+        mutatedSnapshotNumber <- arbitrarySizedNatural `suchThat` (> healthyCloseSnapshotNumber)
```
I've found this was changed in commit [3523b962f5](https://github.com/input-output-hk/hydra/commit/3523b962f505ccc5cbeeb0e2aece26c0c2296666):

> Simplify arbitrary condition to also be consistent with similar mutation for contest

I would add some explanations to make it clear why only checking it's bigger than `healthyCloseSnapshotNumber` is equivalent to `n /= healthySnapshotNumber && n > 0`

ffakenz replied:
I did it so it is consistent with Contest mutations.

It is not fully equivalent, as there are missing values not being tested btw 0 and healthySnapshotNumber but I do believe it is enough to test bigger snapshots than the healthy in order to check this property.

Hopefully this answer your question :)

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -275,10 +325,18 @@ genCloseMutation (tx, _utxo) =
-    , SomeMutation Nothing MutateRequiredSigner <$> do
-        newSigner <- verificationKeyHash <$> genVerificationKey
+    , SomeMutation (Just $ toErrorCode SignerIsNotAParticipant) MutateRequiredSigner <$> do
+        newSigner <- verificationKeyHash <$> genVerificationKey `suchThat` (/= somePartyCardanoVerificationKey)
```
I've found that this was changed in commit [4eaafec002](https://github.com/input-output-hk/hydra/commit/4eaafec002):

> Add first round of label descriptions for close mutations

I would split this commit to only include label descriptions and put this change in a dedicated commit which would explain why it is ok to add `suchThat (/= somePartyCardanoVerificationKey)`

ffakenz replied:
this is not actually needed, as the chances to have this test produce a verification key equal to the healthy one is very unlikely.

anyhow, I added this condition to improve the mutation readability and make it explicit what is the intention.

do you like it or do you prefer to get rid of this?

pgrange replied:
`genVerificationKey suchThat (/= somePartyCardanoVerificationKey)` makes total sense to me: we generate a verification key such that it is not some party verification key.

I just noticed that it was a change with previous test code and wasn't sure how legit the change was from just reading the diff and the commit message.

But now I understand that you actually fixed a bug in the former test were we could have it fail because it accidentally generated **the** incorrect value for this test. Although very improbable your fix makes sense.

Thank you @ffakenz

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -275,10 +325,18 @@ genCloseMutation (tx, _utxo) =
-    , SomeMutation Nothing MutateCloseUTxOHash . ChangeOutput 0 <$> mutateCloseUTxOHash
+    , SomeMutation (Just $ toErrorCode NoSigners) MutateNoRequiredSigner <$> do
+        pure $ ChangeRequiredSigners []
+    , SomeMutation (Just $ toErrorCode TooManySigners) MutateMultipleRequiredSigner <$> do
```
I would explain why it is a problem to have more than one signer for the transaction.

v0d1ch replied:
already answered here https://github.com/input-output-hk/hydra/pull/719#discussion_r1129632155

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -293,15 +351,25 @@ genCloseMutation (tx, _utxo) =
-    , SomeMutation Nothing MutateHeadId <$> do
+    , -- XXX: This is a bit confusing and not giving much value. Maybe we can remove this.
+      -- This also seems to be covered by MutateRequiredSigner
+      SomeMutation (Just $ toErrorCode SignerIsNotAParticipant) CloseFromDifferentHead <$> do
```
❤️ love that mutation name

pgrange reviewed (approved):
Ask for clarifications if my comments are not clear.
If you disagree with my comments, just ignore them and don’t answer or explain.
Hence I, in advance, approve the P.R.

What I like about this P.R.:
* The effort made to explain the intent of the existing mutations
* Some of the simplifications introduced

For me to find it perfect you would have to:
* reduce the size of the P.R. so that we have a simple and clear commit history to walk through (I can't pretend I've reviewed everything)
* see inline comments

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -190,38 +199,77 @@ healthySignature number =
+  | -- | Invalidates the tx by changing the contest snapshot number too old.
+    --
+    -- This is achieved by updating the head input datum to be older, so the
+    -- healthy snapshot number becomes to old.
```
```suggestion
    -- healthy snapshot number becomes too old.
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -190,38 +199,77 @@ healthySignature number =
     MutateToNonNewerSnapshot
-  | -- | Ensure that it's performed by a Head party
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
+    -- used on the tx to be not one of PTs.
```
```suggestion
  | -- | Ensures close is authenticated by one of the Head members by changing the
    -- signer used on the tx to not be one of PTs.
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -190,38 +199,77 @@ healthySignature number =
     MutateRequiredSigner
-  | -- | Ensure output state is consistent with redeemer
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
+    -- used on the tx to be empty.
```
```suggestion
  | -- | Ensures close is authenticated by  one of the Head members by changing the
    -- signer used on the tx to be empty.
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -190,38 +199,77 @@ healthySignature number =
+  | -- | Ensures close is authenticated by a single Head party by changing the signer
+    -- used on the tx to have multiple signers (including the signer to not fail for
+    -- SignerIsNotAParticipant).
+    MutateMultipleRequiredSigner
```
@ch1bo: would you explain why it would be a problem to have more than one signer for the transaction?

pgrange replied:
See https://github.com/input-output-hk/hydra/pull/719#discussion_r1129632155

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -190,38 +199,77 @@ healthySignature number =
+  | -- | Ensures a the signer needs to be added to the head output datum.
     MutateContesters
-  | -- | See spec: 5.5. rule 6 -> value is preserved
+  | -- | Invalidates the tx by changing the output values arbitrarly to be
```
```suggestion
  | -- | Invalidates the tx by changing the output values arbitrarily to be
```

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -230,14 +278,17 @@ genContestMutation
+      , SomeMutation (Just $ toErrorCode SignatureVerificationFailed) MutateSnapshotNumberButNotSignature <$> do
+          mutatedSnapshotNumber <- arbitrarySizedNatural `suchThat` (> healthyContestSnapshotNumber)
+          pure $ ChangeOutput 0 $ changeHeadOutputDatum (replaceSnapshotNumber $ toInteger mutatedSnapshotNumber) headTxOut
+      , SomeMutation (Just $ toErrorCode TooOldSnapshot) MutateToNonNewerSnapshot <$> do
+          mutatedSnapshotNumber <- choose (toInteger healthyContestSnapshotNumber, toInteger healthyContestSnapshotNumber + 1)
```
The following code would pick either 4 or 5:
```Haskell
choose (toInteger healthyContestSnapshotNumber, toInteger healthyContestSnapshotNumber + 1)
```

I would make the code more explicit to explain why `healthyContestSnapshotNumber + 1` is, indeed, a too old snapshot number. I don't understand why by just reading this code.

ffakenz replied:
This change was applied on this [commit](https://github.com/input-output-hk/hydra/commit/3de44f2516569b7ecf6953d6fdc963ebd201b9cf).

There is an explanation about why this was done.

Do you want me to add the commit description as a comment above the code?

pgrange replied:
Thank you @ffakenz for the reference to this commit message. I appreciate that you have a dedicated commit for that with an explanation about it 👍

I must admit that this is a bit tricky case and I'm having a hard time understanding why it's ok. Maybe you just explain it to me live later :)

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -312,6 +380,19 @@ genContestMutation
+      , SomeMutation (Just $ toErrorCode ChangedParameters) MutatePartiesInOutput <$> do
+          mutatedParties <-
+            -- XXX: we need to length of mutatedParties is the same as healthyOnChainParties
+            -- so to not fail because of `must not push contestation deadline`.
```
```suggestion
            -- The length of mutatedParties must be the same as
            --  healthyOnChainParties so to not fail because of
            -- `must not push contestation deadline`.
```

pgrange reviewed (approved):
Last round of inline comments

v0d1ch commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -323,20 +391,15 @@ genCloseMutation (tx, _utxo) =
--- The nice thing to do would probably to generate either "normal" healthy close or
--- or initial healthy close and apply all the mutation to it but we did'nt manage to do that
+-- The nice thing to do would probably to generate either "normal" healthyCloseTx or
+-- or healthyCloseInitialTx and apply all the mutation to it but we did'nt manage to do that
```
```suggestion
-- or healthyCloseInitialTx and apply all the mutations to it but we didn't manage to do that
```

v0d1ch commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -323,20 +391,15 @@ genCloseMutation (tx, _utxo) =
--- or initial healthy close and apply all the mutation to it but we did'nt manage to do that
+-- The nice thing to do would probably to generate either "normal" healthyCloseTx or
+-- or healthyCloseInitialTx and apply all the mutation to it but we did'nt manage to do that
 -- rightaway.
```
```suggestion
-- right away.
```

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -389,6 +390,10 @@ checkContest ctx contestationDeadline contestationPeriod parties closedSnapshotN
     && checkSignedParticipantContestOnlyOnce
     && mustBeWithinContestationPeriod
     && mustUpdateContesters
+    -- XXX: This check is redundant and can be removed,
```
Why don't we remove it in this PR then? Or you wanted to make a separate one just for these XXX items?

ffakenz replied:
Yes we want to avoid changing the on chain scripts in this PR

v0d1ch reviewed (approved):
LGTM! Not sure about all the `XXX` comments though? Perhaps we want to tackle them in a separate PR?

ffakenz committed:
Improve mutation descriptions as per review

ffakenz committed:
Improve mutation readability and make explicit its intention.

Chances to have this test produce a verification key equal to the
healthy one is very unlikely.
Anyhow, the changes make the mutation more readable and explicit,
thus the reason for this change.
This mutation also exists in Close so this also brings consistency
to the table between the two.
